File: lib/tool-jvm/mybatis-auto-wrapper/src/main/java/com/addzero/mybatis/auto_wrapper/AutoWhereUtil.java
package com.addzero.mybatis.auto_wrapper;

import cn.hutool.core.util.ObjUtil;
import cn.hutool.core.util.StrUtil;
import com.baomidou.mybatisplus.core.conditions.AbstractWrapper;
import com.baomidou.mybatisplus.core.conditions.query.LambdaQueryWrapper;
import com.baomidou.mybatisplus.core.conditions.query.QueryWrapper;
import com.baomidou.mybatisplus.core.toolkit.StringUtils;
import com.baomidou.mybatisplus.core.toolkit.Wrappers;

import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.function.BiFunction;

public class AutoWhereUtil {

    @SuppressWarnings({"rawtypes"})
    private static final BiFunction columnProcessToSFunction = (clazz, column) -> {
        Method method = CreateSFunctionUtil.findMethodByColumnName((Class<?>) clazz, (String) column);
        return CreateSFunctionUtil.createSFunction((Class<?>) clazz, method);
    };

    @SuppressWarnings({"rawtypes"})
    private static final BiFunction columnProcessToString = (clazz, column) -> StringUtils.camelToUnderline((String) column);

    @SuppressWarnings({"rawtypes"})
    private static final BiFunction packFieldPostProcessEqField = (entityClass, dtoClazz) -> {
        final Map<String, Field> entityClassMap = getFieldsList((Class<?>) entityClass);
        final Map<String, Field> dtoClazzMap = getFieldsList((Class<?>) dtoClazz);

        dtoClazzMap.entrySet().removeIf(next -> !entityClassMap.containsKey(next.getKey()));
        return new HashSet<>(dtoClazzMap.values());
    };


    private static final BiFunction packFieldPostProcessEqFieldIgnoreId = (entityClass, dtoClazz) -> {
        final Map<String, Field> entityClassMap = getFieldsList((Class<?>) entityClass);
        final Map<String, Field> dtoClazzMap = getFieldsList((Class<?>) dtoClazz);

        dtoClazzMap.entrySet().removeIf(next -> {
            boolean b = !entityClassMap.containsKey(next.getKey());
            return b || StrUtil.equals(next.getKey(), "id");
        });
        return new HashSet<>(dtoClazzMap.values());
    };


    @SuppressWarnings({"rawtypes"})
    private static final BiFunction packFieldPostProcessEmpty = (entityClass, dtoClazz) -> new HashSet<>();

    private static Map<String, Field> getFieldsList(Class<?> clazz) {
        return getFieldsListTailRec(clazz, new HashMap<>());
    }

    //private static Map<String, Field> getFieldsListTailRec(Class<?> clazz, Map<String, Field> map) {
    //    if (clazz == null) return map;
    //    final Field[] declaredFields = clazz.getDeclaredFields();
    //    for (Field field : declaredFields) {
    //        map.putIfAbsent(field.getName(), field);
    //    }
    //    return getFieldsListTailRec(clazz.getSuperclass(), map);
    //}

    private static Map<String, Field> getFieldsListTailRec(Class<?> clazz, Map<String, Field> map) {
        if (clazz == null) return map;
        final Field[] declaredFields = clazz.getDeclaredFields();
        for (Field field : declaredFields) {
            if (!Modifier.isStatic(field.getModifiers())) {
                map.putIfAbsent(field.getName(), field);
            }
        }
        if (map.isEmpty()) throw new RuntimeException("没有找到合适的属性");
        return getFieldsListTailRec(clazz.getSuperclass(), map);
    }

    public static <T> LambdaQueryWrapper<T> lambdaQueryByAnnotation(Class<T> clazz, Object dto) {
        return action(Wrappers.lambdaQuery(), clazz, packFieldPostProcessEmpty, columnProcessToSFunction, dto);
    }

    @SuppressWarnings({"rawtypes", "unchecked"})
    private static <T, W extends AbstractWrapper> W action(W wrapper, Class<T> clazz, BiFunction packFieldPostProcess, BiFunction columnProcess, Object dto) {
        if (ObjUtil.isNull(dto)) {
            return wrapper;
        }

        final GroupInfoBuilder<T, String> groupInfoBuilder = new GroupInfoBuilder<>(clazz, packFieldPostProcess, columnProcess);
        final List<WheresGroupInfo<T, String>> wheresGroupInfos = groupInfoBuilder.buildGroupInfo(dto);
        goTo(clazz, wrapper, wheresGroupInfos);
        return wrapper;
    }

    private static <T, R> void goTo(Class<T> clazz, AbstractWrapper<T, R, ?> wrapper, List<WheresGroupInfo<T, R>> wheresGroupInfos) {
        for (WheresGroupInfo<T, R> wheresGroupInfo : wheresGroupInfos) {
            wheresGroupInfo.process(clazz, wrapper);
        }
    }

    public static <T> LambdaQueryWrapper<T> lambdaQueryByField(Class<T> clazz, Object dto) {
        return action(Wrappers.lambdaQuery(), clazz, packFieldPostProcessEqField, columnProcessToSFunction, dto);
    }

    public static <T> LambdaQueryWrapper<T> lambdaQueryByField(Class<T> clazz, Object dto, boolean ignoreId) {
        return ignoreId ? action(Wrappers.lambdaQuery(), clazz, packFieldPostProcessEqFieldIgnoreId, columnProcessToSFunction, dto) :
                action(Wrappers.lambdaQuery(), clazz, packFieldPostProcessEqField, columnProcessToSFunction, dto);
    }


    public static <T> QueryWrapper<T> queryByAnnotation(Class<T> clazz, Object dto) {
        return action(Wrappers.query(), clazz, packFieldPostProcessEmpty, columnProcessToString, dto);
    }

    public static <T> QueryWrapper<T> queryByField(Class<T> clazz, Object dto) {
        return action(Wrappers.query(), clazz, packFieldPostProcessEqField, columnProcessToString, dto);
    }

}

File: lib/tool-jvm/database/mybatis-auto-wrapper/src/main/kotlin/site/addzero/mybatis/auto_wrapper/AutoWhereUtil.kt
@file:JvmName("AutoWhereUtil")
package site.addzero.mybatis.auto_wrapper
import cn.hutool.core.util.ObjUtil
import cn.hutool.core.util.ReflectUtil
import cn.hutool.core.util.StrUtil
import com.baomidou.mybatisplus.core.conditions.AbstractWrapper
import com.baomidou.mybatisplus.core.conditions.query.LambdaQueryWrapper
import com.baomidou.mybatisplus.core.conditions.query.QueryWrapper
import com.baomidou.mybatisplus.core.toolkit.StringUtils
import com.baomidou.mybatisplus.core.toolkit.Wrappers
import com.baomidou.mybatisplus.core.toolkit.support.SFunction
import java.lang.reflect.Field

@Suppress("UNCHECKED_CAST")
private val columnProcessToSFunction: (Class<*>, String) -> SFunction<*, *> = { clazz, column ->
    val method = CreateSFunctionUtil.findMethodByColumnName(clazz, column)
    CreateSFunctionUtil.createSFunction(clazz, method)!!
}
private val columnProcessToString: (Class<*>, String) -> String = { _, column -> StringUtils.camelToUnderline(column)!! }
private val packFieldPostProcessEqField: (Class<*>, Class<*>) -> MutableSet<Field> = { entityClass, dtoClazz ->
    val entityClassMap = getFieldsList(entityClass)
    val dtoClazzMap = getFieldsList(dtoClazz)
    dtoClazzMap.filterKeys { entityClassMap.containsKey(it) }.values.toMutableSet()
}
private val packFieldPostProcessEqFieldIgnoreId: (Class<*>, Class<*>) -> MutableSet<Field> = { entityClass, dtoClazz ->
    val entityClassMap = getFieldsList(entityClass)
    val dtoClazzMap = getFieldsList(dtoClazz)

    dtoClazzMap.filterKeys {
        entityClassMap.containsKey(it) && !StrUtil.equals(it, "id")
    }.values.toMutableSet()
}
private val packFieldPostProcessEmpty: (Class<*>, Class<*>) -> MutableSet<Field> = { _, _ -> mutableSetOf() }
private fun getFieldsList(clazz: Class<*>): Map<String, Field> {
    return generateSequence(clazz) { it.superclass }.flatMap { ReflectUtil.getFields(it).asSequence() }.associateBy { it.name }
}

fun <T> lambdaQueryByAnnotation(clazz: Class<T>, dto: Any): LambdaQueryWrapper<T> {
    @Suppress("UNCHECKED_CAST") val columnProcess = columnProcessToSFunction as (Class<T>, String) -> SFunction<T, *>
    return action(Wrappers.lambdaQuery<T>(), clazz, packFieldPostProcessEmpty, columnProcess, dto)
}

private fun <T, R, W : AbstractWrapper<T, R, *>> action(wrapper: W, clazz: Class<T>, packFieldPostProcess: (Class<*>, Class<*>) -> MutableSet<Field>, columnProcess: (Class<T>, String) -> R, dto: Any?): W {
    if (ObjUtil.isNull(dto)) {
        return wrapper
    }

    val groupInfoBuilder = GroupInfoBuilder(clazz, packFieldPostProcess, columnProcess)
    val wheresGroupInfos = groupInfoBuilder.buildGroupInfo(dto!!)
    goTo(clazz, wrapper as AbstractWrapper<T, R, *>, wheresGroupInfos)
    return wrapper
}

private fun <T, R> goTo(clazz: Class<T>, wrapper: AbstractWrapper<T, R, *>, wheresGroupInfos: List<WheresGroupInfo<T, R>>) {
    wheresGroupInfos.forEach { wheresGroupInfo ->
        val joinAndNested = wheresGroupInfo as JoinAndNested<T, R>
        joinAndNested.process(clazz, wrapper)
    }
}

@Suppress("UNCHECKED_CAST")
fun <T> lambdaQueryByField(clazz: Class<T>, dto: Any): LambdaQueryWrapper<T> {
    val columnProcess = columnProcessToSFunction as (Class<T>, String) -> SFunction<T, *>
    return action(Wrappers.lambdaQuery<T>(), clazz, packFieldPostProcessEqField, columnProcess, dto)!!
}

@Suppress("UNCHECKED_CAST")
fun <T> lambdaQueryByField(clazz: Class<T>, dto: Any, ignoreId: Boolean): LambdaQueryWrapper<T> {
    val columnProcess = columnProcessToSFunction as (Class<T>, String) -> SFunction<T, *>
    return (if (ignoreId) action(Wrappers.lambdaQuery<T>(), clazz, packFieldPostProcessEqFieldIgnoreId, columnProcess, dto) else action(Wrappers.lambdaQuery<T>(), clazz, packFieldPostProcessEqField, columnProcess, dto))!!
}


fun <T> queryByAnnotation(clazz: Class<T>, dto: Any): QueryWrapper<T> {
    return action(Wrappers.query<T>(), clazz, packFieldPostProcessEmpty, columnProcessToString, dto)!!
}

fun <T> queryByField(clazz: Class<T>, dto: Any): QueryWrapper<T> {
    return action(Wrappers.query<T>(), clazz, packFieldPostProcessEqField, columnProcessToString, dto)!!
}
